ghcide/src/Development/IDE/Core/ProgressReporting.hs

```diff
@@ -0,0 +1,229 @@
+import           Data.HashMap.Strict            (HashMap)
+import qualified Data.HashMap.Strict            as HMap
+import           Data.IORef
+import qualified Data.Text                      as T
+import           Data.Unique
+import           Development.IDE.GHC.Orphans    ()
+import           Development.IDE.Graph          hiding (ShakeValue)
+import           Development.IDE.Types.Location
+import           Development.IDE.Types.Options
+import           GHC.IORef                      (atomicModifyIORef'_,
+                                                 atomicSwapIORef)
+import qualified Language.LSP.Server            as LSP
+import           Language.LSP.Types
+import qualified Language.LSP.Types             as LSP
+import           System.Time.Extra
+import           UnliftIO.Exception             (bracket_)
+
+data ProgressEvent
+    = KickStarted
+    | KickCompleted
+
+data ProgressReporting  = ProgressReporting
+  { progressUpdate :: ProgressEvent -> IO ()
+  , inProgress     :: forall a. NormalizedFilePath -> Action a -> Action a
+  , progressStop   :: IO ()
+  }
+
+noProgressReporting :: IO ProgressReporting
+noProgressReporting = return $ ProgressReporting
+  { progressUpdate = const $ pure ()
+  , inProgress = const id
+  , progressStop   = pure ()
+  }
+
+-- | A 'ProgressReporting' that sends the WorkDone Begin and End notifications
+--   synchronously. Progress notifications are sent from a sampling thread.
+directProgressReporting
+    :: Double -- ^ sampling rate
+    -> Maybe (LSP.LanguageContextEnv config)
+    -> ProgressReportingStyle
+    -> IO ProgressReporting
+directProgressReporting sample env style = do
+    st <- newIORef Nothing
+    inProgressVar <- newIORef (HMap.empty @NormalizedFilePath @Int)
+
+    let progressUpdate KickStarted = do
+          u <- newProgressToken
+          writeIORef st (Just u)
+          mRunLspT env $ start u
+        progressUpdate KickCompleted = do
+            mbToken <- atomicSwapIORef st Nothing
+            for_ mbToken $ \u ->
+                mRunLspT env $ stop u
+
+        inProgress file = actionBracket (f file succ) (const $ f file pred) . const
+        -- This function is deliberately eta-expanded to avoid space leaks.
+        -- Do not remove the eta-expansion without profiling a session with at
+        -- least 1000 modifications.
+        f file shift = atomicModifyIORef'_ inProgressVar $
+                HMap.insertWith (\_ x -> shift x) file (shift 0)
+
+        progressLoop :: Double -> LSP.LspM a ()
+        progressLoop prev = do
+            mbToken <- liftIO $ readIORef st
+            case mbToken of
+                Nothing ->
+                    liftIO (sleep sample) >> progressLoop 0
+                Just t -> do
+                    current <- liftIO $ readIORef inProgressVar
+                    prev <- progress style prev current t
+                    liftIO $ sleep sample
+                    progressLoop prev
```

Comment:
It might be clearer if you calculated the value inside the case, then had the `sleep` and `processLoop` at the end of the case, so its easier to see the control flow is just an uninterupted loop.